lib/jwe.js

```diff
@@ -0,0 +1,511 @@
+      var result = decryptCEKHelper(header.alg, encrypted_cek, jweKeyStore[i], log);
+      if (!result.error && result.cek)
+        return result;
+    }
+  }
+
+  return { 'error': new Error('tried all keys to decryt cek but none of them works'), 'cek': null };
+};
+
+/**
+ * decrypt content function
+ *
+ * @param{Object}  header: header
+ * @param{Buffer}  cek: cek
+ * @param{Buffer}  cipherText: encrypted content
+ * @param{Buffer}  iv: initialization vector
+ * @param{Buffer}  authTag: authentication tag
+ * @param{Buffer}  aad: additional authentication information
+ * @return{Function}  log: logging function
+ */
+var decrytContent = (header, cek, cipherText, iv, authTag, aad, log) => {
+  if (!header.enc)
+    return { 'error': new Error('enc is missing in JWE header'), 'content': null };
+  if (['A128GCM', 'A256GCM', 'A128CBC-HS256', 'A192CBC-HS384', 'A256CBC-HS512'].indexOf(header.enc) === -1)
+    return { 'error' : new Error('Unsupported enc in JWE header: ' + header.enc), 'content': null };
+
+  var mapper = { 
+    'A128GCM': 'aes-128-gcm', 
+    'A256GCM': 'aes-256-gcm', 
+    'A128CBC-HS256': 'aes-128-cbc-hmac-sha-256',
+    'A192CBC-HS384': 'aes-192-cbc-hmac-sha-384',
+    'A256CBC-HS512': 'aes-256-cbc-hmac-sha-512'
+  };
+
+  try {
+    var content = null;
+
+    if (header.enc === 'A128GCM' || header.enc === 'A256GCM') {
+      var decipher = crypto.createDecipheriv(mapper[header.enc], cek, iv);
+      decipher.setAAD(aad);
+      decipher.setAuthTag(authTag);
+      content = decipher.update(cipherText);
+      content = Buffer.concat([content, decipher.final()]);
+    } else {
+      content = decrypt_AES_CBC_HMAC_SHA2(mapper[header.enc], cek, cipherText, iv, aad, authTag);    
+    }
+
+    return { 'error': null, 'content': content.toString() };
+  } catch (ex) {
+    return { 'error': ex, 'content': null };
+  }
+};
+
+/**
+ * decrypt content function for the case where alg is dir
+ *
+ * @param{Object}  header: header
+ * @param{Buffer}  jweKeyStore: list of keys
+ * @param{Buffer}  cipherText: encrypted content
+ * @param{Buffer}  iv: initialization vector
+ * @param{Buffer}  authTag: authentication tag
+ * @param{Buffer}  aad: additional authentication information
+ * @return{Function}  log: logging function
+ */
+var decrytContentForDir = (header, jweKeyStore, cipherText, iv, authTag, aad, log) => {
+  var key = null;
+
+  // try the key with the corresponding kid
+  if (header.kid) {
+    for (var i = 0; i < jweKeyStore.length; i++) {
+      if (header.kid === jweKeyStore[i].kid && jweKeyStore[i].kty === 'oct') {
+        key = jweKeyStore[i];
+        log.info('found a key matching kid: ' + header.kid);
+        break;
+      }
+    }
+
+    if (key)
+      return decrytContent(header, base64url.toBuffer(key.k), cipherText, iv, authTag, aad, log);
+  }
+
+  log.info('In decrytContentForDir: we cannout find a key using kid, now we try all the possible keys');
```

Comment:
similar to above, consider trying all keys if kid is null. If kid is not matched, then most likely a key refresh is needed.